keps/sig-windows/1981-windows-privileged-container-support/README.md

```diff
@@ -508,27 +508,64 @@ When a new `hostProcess` container is created with the name of a local users Gro
     Unix domain sockets can be accessed via their paths on the host like named pipes.
-  - Mounting directories from the host OS into `hostProcess` containers will work just like with normal containers but this is not recommend.
-    Instead workloads should access the host OS's file-system as if it was not being run in a container.
+- Mounting directories from the host OS into `hostProcess` containers will work just like with normal containers but this is not recommend.
+    Instead workloads should access the host OS's file-system as if not being run in a container.
   - All other volume types supported for normal containers on Windows will work with `hostProcess` containers.
 - `HostProcess` Containers will have full access to the host file-system (unless restricted by filed-based ACLs and the run_as_username used to start the container).
 - There will be no `chroot` equivalent.
 
-- Note: Behavior of volume mounts will differ between the alpha/beta (old) implementation of this feature and the stable (new) implementation.
-Designing/testing/validation of an acceptable solution for handling volume mounts w.r.t. `hostProcess` containers was a primary reason for keeping the featuer in `beta`. (Previous behavior volume mount behavior can be viewed [here](https://github.com/kubernetes/enhancements/blob/cc4052fa88508fcbb20b978b8c181d06acf734b1/keps/sig-windows/1981-windows-privileged-container-support/README.md#container-mounts))
-A recording of the behaviors differces from a SIG-Windows community meeting can be found [here](https://youtu.be/8GeZKXgvkdY?t=309).
-  - Also note -  In the recording it was mentioned that this functionally might not be supported on WS2019. This functionality will be avaible in WS2019 but will require an OS patch (ETA: July 2022).
-- Containerd v1.7+ will be required for new volume mount behavior.
+##### Compatibility
+
+During the alpha/beta implementations of this feature only **Symlink** volume mount behavior was implmented.
+This implemention did unlock a lot of criticle use cases for managing Windows nodes in Kubernets clusters but did have some usability issues
+(such as https://pkg.go.dev/k8s.io/client-go/rest#InClusterConfig not working as expected).
+
+The **bind mount** volume mount behavior gives full access to the host OS's filesystem (an explicit goal of this enhancement) and addreses the usability issues with the initial approach.
+This approach requires the use of Windows OS APIs that were not present in Windows Server 2019 during alpha/beta implmentations of this feature.
+These APIs *will* be available in WS2019 beginning in July 2022 with the monthly OS security patches.
+Containerd v1.7+ will be required for this behavior.
+
+In order to maintain backwards compatility with container images that were built from around the **symlink** volume mount behavior an annotation can be used to specify which volume mounting behavior should be used.
```

Comment:
Instead of needing an annotation, can we set `$CONTAINER_SANDBOX_MOUNT_POINT` regardless of which method was used?

Reply:
We still need an annotation to have the container runtime use different mount behaviors.

For example,
with the `symlink` behavior a volume mount with a mount path of `/mnt/foo` would should up at `c:\C\
{containerid}\mnt\foo` 

But with `bind` behavior the volume would with a month path of `/mnt/foo` would show up at `c:\mnt\foo`

For the `bind` behavior `$CONTATINER_SANDBOX_MOUNT_POINT` will be set to `c:\hpc` (which is where the containers of the container image get mounted to)

Reply:
I can add these examples to the KEP in my next set up updates

Reply:
In most cases, we are able to automatically detect the correct mount behavior to use. The only scenario is when both methods are available and there's a choice. Is there a reason why someone would not choose the bind mount method if available?

Annotations are considered part of the API, and since this is a pod API which is v1, that means if we add the annotation we'll need to support this forever according to the [deprecation policy.](https://kubernetes.io/docs/reference/using-api/deprecation-policy/)

Reply:
If it would be desirable to not add an annotation to switch behaviors we can instead try an approach like 
- If you are on containerd v1.6.x you'll get the `symlink` behavior
- If you are on containerd v1.7+ you'll get the `bind` behavior. 
Containerd and below handle configuring the bind-mounts so this is another approach (a bit less configurable for end-users but maybe cleaner in the long run?)

Reply:
@liggitt can you comment on this?

> Annotations are considered part of the API, and since this is a pod API which is v1, that means if we add the annotation we'll need to support this forever according to the [deprecation policy.](https://kubernetes.io/docs/reference/using-api/deprecation-policy/)

My understanding is that annotations exist so that we can have flexible behaviors in stable APIs without needing to persist potentially temporary fields in stable APIs (which would require forever support).

Reply:
Ah i see 

> The following rules govern the deprecation of elements of the API. This includes:
> - REST resources (aka API objects)
> - Fields of REST resources
> - Annotations on REST resources, including "beta" annotations but not including "alpha" annotations.
> - Enumerated or constant values
> - Component config structures

Thanks for the info!

Reply:
I don't have any concerns. Today we have code that checks if the relevant functionality we need for bind mounts to work is available, and if not we'll fallback to the symlink behavior. Once the ws2019 patch is out with this support, we can probably axe that and error out and have bind be the only thing supported on Containerd 1.7+. Removing the annotations approach would be lovely though haha

Reply:
> containerd v1.7x will always use the bind behavior (we can fail with easy to consume errors if the required APIs are not available).

Does this mean we need to tie containerd v1.7 with the newer Windows verssion too?

> This way we can users who depend on the symlink volume mount behaviors can stay on containerd v1.6 until all their workloads have been updated.

Since we set the env variable in both `symlink` and `bind` modes, users may not need to update their workloads right?

Reply:
The environment variable will still be set, however we don't bind the mounts where they used to be placed with the symlink behavior so the environment variable loses some of its use. Previously if you asked a mount to show up at C:\path\to\foo, it would get symlinked to C:\C\{containerid}\path\to\foo, and users could access via %CONTAINER_SANDBOX_MOUNT_POINT%\path\to\foo. Now that we have a way to have these be unique per container, we just bind them exactly where they were asked to be placed, so C:\path\to\foo which is definitely improved behavior I'd say but accessing the mount path via the env var won't work anymore.

We could probably find a way to additionally bind the mount twice, once to where it was asked, and once to the old symlink location, however I'll need to see if this is feasible.. The rootfs for the container in the new approach is a bind mount as well to (by default) C:\hpc, and there seems to be some odd behavior with nesting these.

Reply:
> Does this mean we need to tie containerd v1.7 with the newer Windows version too?

For running host-process containers on Windows Server 2019 yes.
On Windows users should always be installing the latest security patches (which would include the needed APIs) so I do not think this will be an issue.

Windows Server 2022 shipped with the required APIs so the `bind` behavior will work on any set of OS patches.

Reply:
> * containerd v1.6.x will always use the `symlink` behavior (any other behaviors for v1.6.x releases would need containerd maintainers to approve)
> * containerd v1.7x will always use the `bind` behavior (we can fail with easy to consume errors if the required APIs are not available).

This sounds fine to me. This might be even better since it will force users to adopt the long-term solution.  We will need to document the upgrade path and what changes are required when moving to contianerd 1.7.

Reply:
> users could access via %CONTAINER_SANDBOX_MOUNT_POINT%\path\to\foo. Now that we have a way to have these be unique per container, we just bind them exactly where they were asked to be placed, so C:\path\to\foo

What I was wondering from a backwards compatibility perspective, is if we could set `$CONTAINER_SANDBOX_MOUNT_POINT` to something like `C:` in bind mode, so that users could have some time to transition to the new behavior.

Reply:
We can continue brainstorming on this, but I want to see if we can find a way to support users upgrading from the symlink to bind mount methods in a non-disruptive way.

One approach that was discussed earlier was to have an annotation for users to opt-in to the new behavior.  And we could deprecate and eventually remove the old behavior. From the k8s deprecation policy, the time period for deprecation is [one year](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-feature-or-behavior). One downside to this approach is that we need to supoprt both modes in the runtime, and that it's not easy to detect if users are depending on the old behavior. Despite ample warning about deprecations, users may not be able to keep up and get caught by surprise when their workloads break on upgrade.

Another idea I contemplated is to introduce a second env variable `CONTAINER_VOLUME_ROOT` that points to the same directory in containerd 1.6, but then points to the new directory in containerd 1.7. That way even when upgrading to the newer containerd version, the workloads will continue to work. We would still need to tell users to switch to the new env variable, so it does require work from them to avoid being broken, but at least this could help support a non-disruptive upgrade sequence in the future.

Reply:
@msau42 I'm prototyping an approach to keep things backwards compatible. The gist would be we'd bind mounts where the user asked, but in addition we'll also bind them to a relative path under the rootfs like they were in the symlink approach. We still set the env var and have it point to C:\hpc as described so $env:CONTAINER_SANDBOX_MOUNT_POINT\path\to\mount should still work.

Reply:
Reviving this to ask, currently Calico is using [this relatively convoluted method](https://github.com/kubernetes-sigs/sig-windows-tools/blob/master/hostprocess/calico/install/calico-install.ps1#L19-L22) to build a kubeconfig based off service account (cc: @marosset / @jsturtevant) and we're doing the same for Windows NPM, is there a migration story for these scenarios? Is it possible to make clientgo aware of this change to be able to use rest.InClusterConfig()?